In [1]:
pip install sqlalchemy pymysql pandas matplotlib seaborn python-dotenv

Note: you may need to restart the kernel to use updated packages.


In [2]:
import datetime
import requests
import json
import os

from dotenv import load_dotenv
import pandas as pd
import sqlalchemy as sa
from sqlalchemy import create_engine, text, MetaData, Table
from sqlalchemy.orm import sessionmaker

In [3]:
def create_connection():
    """
    Створює підключення через SQLAlchemy
    """
    # Завантажуємо змінні середовища
    load_dotenv()

    
    host = os.getenv('DB_HOST', 'localhost')
    port = os.getenv('DB_PORT', '3306')
    user = os.getenv('DB_USER')
    password = os.getenv('DB_PASSWORD')
    database = os.getenv('DB_NAME')

    if not all([user, password, database]):
        raise ValueError("Не всі параметри БД задані в .env файлі!")


    connection_string = f"mysql+pymysql://{user}:{password}@{host}:{port}/{database}"


    engine = create_engine(
        connection_string,
        pool_size=2,           # Розмір пулу підключень
        max_overflow=20,        # Максимальна кількість додаткових підключень
        pool_pre_ping=True,     # Перевірка підключення перед використанням
        echo=False              # Логування SQL запитів (True для debug)
    )

 # Тестуємо підключення
    try:
        with engine.connect() as conn:
            result = conn.execute(text("SELECT 1"))
            result.fetchone()

        print("✅ Підключення до БД успішне!")
        print(f"🔗 {user}@{host}:{port}/{database}")
        print(f"⚡ Engine: {engine}")

        return engine

    except Exception as e:
        print(f"❌ Помилка підключення: {e}")
        return None

# Створюємо підключення
engine = create_connection()

✅ Підключення до БД успішне!
🔗 root@localhost:3306/classicmodels
⚡ Engine: Engine(mysql+pymysql://root:***@localhost:3306/classicmodels)


# Домашнє завдання: Внесення оновлень в БД і робота з транзакціями

Це ДЗ передбачене під виконання на локальній машині. Виконання з Google Colab буде суттєво ускладнене.

## Підготовка
1. Переконайтесь, що у вас встановлены необхідні бібліотеки:
   ```bash
   pip install sqlalchemy pymysql pandas matplotlib seaborn python-dotenv
   ```

2. Створіть файл `.env` з параметрами підключення до бази даних classicmodels. Базу даних ви можете отримати через

  - docker-контейнер згідно існтрукції в [документі](https://www.notion.so/hannapylieva/Docker-1eb94835849480c9b2e7f5dc22ee4df9), також відео інструкції присутні на платформі - уроки "MySQL бази, клієнт для роботи з БД, Docker і ChatGPT для запитів" та "Як встановити Docker для роботи з базами даних без терміналу"
  - або встановивши локально цю БД - для цього перегляньте урок "Опціонально. Встановлення MySQL та  БД Сlassicmodels локально".
  
  Приклад `.env` файлу ми створювали в лекції. Ось його обовʼязкове наповнення:
    ```
    DB_HOST=your_host
    DB_PORT=3306 або 3307 - той, який Ви налаштували
    DB_USER=your_username
    DB_PASSWORD=your_password
    DB_NAME=classicmodels
    ```
  Якщо ви створили цей файл під час перегляду лекції - **новий створювати не треба**. Замініть лише назву БД, або пропишіть назву в коді створення підключення (замість отримання назви цільової БД зі змінних оточення). Але переконайтесь, що до `.env` файл лежить в тій самій папці, що і цей ноутбук.

  **УВАГА!** НЕ копіюйте скрит для **створення** `.env` файлу. В лекції він наводиться для прикладу. І давалось пояснення, що в реальних проєктах ми НІКОЛИ не пишемо доступи до бази в коді. Копіювання скрипта для створення `.env` файлу сюди в ДЗ буде вважатись грубою помилкою і ми зніматимемо бали.

3. Налаштуйте підключення через SQLAlchemy до БД за прикладом в лекції.

Рекомендую вивести (відобразити) змінну engine після створення. Вона має бути не None! Якщо None - значить у Вас не підтягнулись налаштування з .env файла.

Ви також можете налаштувати параметри підключення до БД без .env файла, просто прописавши текстом в відповідних місцях. Це - не рекомендований підхід.


## Завдання

### Завдання 1: Оновлення інформації про клієнта (2 бали)

**Створіть функцію для оновлення контактної інформації клієнта за його номером** з наступними можливостями:
- Оновлення телефону клієнта
- Оновлення email (якщо поле існує в таблиці)

Опціонально, якщо вам хочеться більше практики:
- Логування змін в окрему таблицю

Використайте підхід з параметризованими запитами через `text()` та `UPDATE` оператор. Не забудьте на початку перевірити чи існує клієнт з таким номером в базі - це хороша практика.

Отримати всі колонки, які існують в таблиці ви можете наступним запитом
```
  SELECT COLUMN_NAME, DATA_TYPE
  FROM INFORMATION_SCHEMA.COLUMNS
  WHERE TABLE_NAME = 'customers'
```

Запустіть функцію і продемонструйте її роботу, запустивши SELECT, який допоможе це зробити.



In [4]:
pd.read_sql(''' SELECT COLUMN_NAME, DATA_TYPE
  FROM INFORMATION_SCHEMA.COLUMNS
  WHERE TABLE_NAME = 'customers'
''',engine)

,COLUMN_NAME,DATA_TYPE
0,customerNumber,int
1,customerName,varchar
2,contactLastName,varchar
3,contactFirstName,varchar
4,phone,varchar
5,addressLine1,varchar
6,addressLine2,varchar
7,city,varchar
8,state,varchar
9,postalCode,varchar


In [5]:
def update_customer_phone(cust_number,new_phone):
    
    with engine.connect() as conn:
        check_query = text("SELECT contactFirstName, contactLastName FROM customers WHERE customerNumber = :cust_number")
        
        result = conn.execute(check_query, {'cust_number':cust_number})
        customer = result.fetchone()
        
        if not customer:
            print(f"❌ Клієнт {cust_number} не знайдений")
            return False
            

        
        conn.execute(
            text('''
                UPDATE customers 
                SET phone = :new_phone
                WHERE customerNumber = :cust_number
            '''),
            {'cust_number':cust_number, 'new_phone':new_phone}
        )
        conn.commit()

In [6]:
update_customer_phone(103, '11.11.1111')
pd.read_sql("SELECT customerNumber, customerName, phone FROM customers WHERE customerNumber = 103",engine)

,customerNumber,customerName,phone
0,103,Atelier graphique,11.11.1111


In [7]:
update_customer_phone(99999, '22.22.2222')

❌ Клієнт 99999 не знайдений


False

### Завдання 2: Створення нового замовлення з транзакцією (5 балів)

**Реалізуйте процес створення нового замовлення** з наступними кроками в одній транзакції:
- Створення запису в таблиці `orders`
- Додавання товарних позицій в `orderdetails`
- Перевірка наявності товарів на складі
- Зменшення кількості товарів на складі

Запустіть процес з тестовими даними і продемонструйте через SELECT, що процес успішно відпрацював і були виконані необхідні операції.




In [8]:
pd.read_sql("SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = 'orders'",engine)

,COLUMN_NAME,DATA_TYPE
0,orderNumber,int
1,orderDate,date
2,requiredDate,date
3,shippedDate,date
4,status,varchar
5,comments,text
6,customerNumber,int


In [9]:
pd.read_sql("SELECT * FROM orders LIMIT 10",engine)

,orderNumber,orderDate,requiredDate,shippedDate,status,comments,customerNumber
0,9000,2026-09-27,2026-10-05,None,In Process,NaN,103
1,10100,2003-01-06,2003-01-13,2003-01-10,Shipped,NaN,363
2,10101,2003-01-09,2003-01-18,2003-01-11,Shipped,Check on availability.,128
3,10102,2003-01-10,2003-01-18,2003-01-14,Shipped,NaN,181
4,10103,2003-01-29,2003-02-07,2003-02-02,Shipped,NaN,121
5,10104,2003-01-31,2003-02-09,2003-02-01,Shipped,NaN,141
6,10105,2003-02-11,2003-02-21,2003-02-12,Shipped,NaN,145
7,10106,2003-02-17,2003-02-24,2003-02-21,Shipped,NaN,278
8,10107,2003-02-24,2003-03-03,2003-02-26,Shipped,Difficult to negotiate with customer. We need ...,131
9,10108,2003-03-03,2003-03-12,2003-03-08,Shipped,NaN,385


In [10]:
pd.read_sql("SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = 'orderdetails'",engine)

,COLUMN_NAME,DATA_TYPE
0,orderNumber,int
1,productCode,varchar
2,quantityOrdered,int
3,priceEach,decimal
4,orderLineNumber,smallint


In [11]:
pd.read_sql("SELECT * FROM orderdetails LIMIT 3",engine)

,orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber
0,10100,S18_1749,30,136.00,3
1,10100,S18_2248,50,55.09,2
2,10100,S18_4409,22,75.46,4


In [12]:
pd.read_sql("SELECT * FROM products LIMIT 10",engine)

,productCode,productName,productLine,productScale,productVendor,productDescription,quantityInStock,buyPrice,MSRP
0,S10_1678,1969 Harley Davidson Ultimate Chopper,Motorcycles,1:10,Min Lin Diecast,"This replica features working kickstand, front...",7933,48.81,95.70
1,S10_1949,1952 Alpine Renault 1300,Classic Cars,1:10,Classic Metal Creations,Turnable front wheels; steering function; deta...,7305,98.58,214.30
2,S10_2016,1996 Moto Guzzi 1100i,Motorcycles,1:10,Highway 66 Mini Classics,"Official Moto Guzzi logos and insignias, saddl...",6625,68.99,118.94
3,S10_4698,2003 Harley-Davidson Eagle Drag Bike,Motorcycles,1:10,Red Start Diecast,"Model features, official Harley Davidson logos...",5582,91.02,193.66
4,S10_4757,1972 Alfa Romeo GTA,Classic Cars,1:10,Motor City Art Classics,Features include: Turnable front wheels; steer...,3252,85.68,136.00
5,S10_4962,1962 LanciaA Delta 16V,Classic Cars,1:10,Second Gear Diecast,Features include: Turnable front wheels; steer...,6791,103.42,147.74
6,S12_1099,1968 Ford Mustang,Classic Cars,1:12,Autoart Studio Design,"Hood, doors and trunk all open to reveal highl...",68,95.34,194.57
7,S12_1108,2001 Ferrari Enzo,Classic Cars,1:12,Second Gear Diecast,Turnable front wheels; steering function; deta...,3619,95.59,207.80
8,S12_1666,1958 Setra Bus,Trucks and Buses,1:12,Welly Diecast Productions,"Model features 30 windows, skylights & glare r...",1579,77.90,136.67
9,S12_2823,2002 Suzuki XREO,Motorcycles,1:12,Unimax Art Galleries,"Official logos and insignias, saddle bags loca...",9997,66.27,150.62


In [13]:
pd.read_sql(
    """
    SELECT *
    FROM products
    WHERE productCode = 'S18_1749'
    """,
    engine
)

,productCode,productName,productLine,productScale,productVendor,productDescription,quantityInStock,buyPrice,MSRP
0,S18_1749,1917 Grand Touring Sedan,Vintage Cars,1:18,Welly Diecast Productions,This 1:18 scale replica of the 1917 Grand Tour...,2717,86.7,170.0


In [14]:
pd.read_sql("SELECT * FROM orderdetails LIMIT 3",engine)

,orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber
0,10100,S18_1749,30,136.00,3
1,10100,S18_2248,50,55.09,2
2,10100,S18_4409,22,75.46,4


In [15]:
def create_order(customer_number, product_code, quantity):
    
    with engine.begin() as conn:
        
        result_order = conn.execute(text("SELECT MAX(orderNumber) + 1 FROM orders"))
        new_order_number = result_order.fetchone()[0]

        check_stock = text("SELECT quantityInStock,MSRP FROM products WHERE productCode = :product_code")
        result = conn.execute(check_stock, {'product_code':product_code})
        result_product = result.fetchone()
        sale_price = result_product[1]
        stock = result_product[0]       
        if stock < quantity:                   
            raise Exception("Товара не хватает")
    
        #Створення запису в таблиці `orders`
        
        conn.execute(
            text('''INSERT INTO orders (orderNumber,orderDate,requiredDate,status,customerNumber) 
            VALUES (:order_number, CURRENT_DATE, DATE_ADD(CURRENT_DATE, INTERVAL 7 DAY), :status, :customer_number)'''),
            {'order_number': new_order_number,
             'status': 'In Process',
             'customer_number': customer_number
            }
        )
        #Додавання товарних позицій в `orderdetails`
        conn.execute(
            text('''INSERT INTO orderdetails(orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)
            VALUES(:order_number, :product_code, :quantity_ordered, :price_each, :order_line_number)'''),
            {'order_number':new_order_number,
             'product_code':product_code,
             'quantity_ordered':quantity,
             'price_each':sale_price,
             'order_line_number':1}
        )
    
        conn.execute(
             text('''
                 UPDATE products
                 SET quantityInStock = quantityInStock - :quantity_ordered
                 WHERE productCode = :product_code'''),
                {'quantity_ordered':quantity,
                 'product_code':product_code}
        )
            
    

In [16]:
create_order(103,'S18_1749',5)

In [17]:
pd.read_sql(
    """
    SELECT *
    FROM orderdetails
    WHERE orderNumber = (SELECT MAX(orderNumber) FROM orders)
    """,
    engine
)

,orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber
0,10428,S18_1749,5,170.0,1


In [18]:
pd.read_sql(
    """
    SELECT *
    FROM products
    WHERE productCode = 'S18_1749'
    """,
    engine
)

,productCode,productName,productLine,productScale,productVendor,productDescription,quantityInStock,buyPrice,MSRP
0,S18_1749,1917 Grand Touring Sedan,Vintage Cars,1:18,Welly Diecast Productions,This 1:18 scale replica of the 1917 Grand Tour...,2712,86.7,170.0


In [22]:
pd.read_sql("SELECT * FROM orders ORDER BY orderNumber DESC LIMIT 10",engine)

,orderNumber,orderDate,requiredDate,shippedDate,status,comments,customerNumber
0,10428,2026-09-27,2026-10-04,None,In Process,NaN,103
1,10427,2026-09-27,2026-10-04,None,In Process,NaN,103
2,10426,2026-09-27,2026-10-04,None,In Process,NaN,103
3,10425,2005-05-31,2005-06-07,None,In Process,NaN,119
4,10424,2005-05-31,2005-06-08,None,In Process,NaN,141
5,10423,2005-05-30,2005-06-05,None,In Process,NaN,314
6,10422,2005-05-30,2005-06-11,None,In Process,NaN,157
7,10421,2005-05-29,2005-06-06,None,In Process,Custom shipping instructions were sent to ware...,124
8,10420,2005-05-29,2005-06-07,None,In Process,NaN,282
9,10419,2005-05-17,2005-05-28,2005-05-19,Shipped,NaN,382


In [23]:
create_order(103,'S18_1749',9999)

Exception: Товара не хватает

Функція create_order(customer_number, product_code, quantity) створює замовлення в одній транзакції (engine.begin()). Номер замовлення обчислюється автоматично (MAX(orderNumber) + 1), ціна береться з каталогу (MSRP), дати встановлює сама база (CURRENT_DATE та +7 днів). У тілі функції немає жорстко вписаних значень, тому її можна викликати для будь-якого клієнта, товару та кількості.
Успішний тест: create_order(103, 'S18_1749', 5) створив замовлення 10428 зі статусом In Process, додав позицію з ціною 170.0 (MSRP), а залишок товару на складі зменшився з 2717 до 2712, тобто рівно на 5 одиниць.
Тест на нестачу товару: виклик із кількістю 9999 при залишку 2712 зупинився з помилкою «Товара не хватает». Після цього нове замовлення не з'явилося (останній номер залишився 10428), а залишок на складі не змінився (2712).
Перевірку наявності товару розміщено до запису в базу, щоб функція зупинялася одразу і не створювала замовлення, яке неможливо виконати. Навіть якби перевірка стояла після запису, транзакція відкотила б усі зміни, і в базі не залишилося б замовлення без списання товару зі складу.
Під час тестування в базі з'явилися три тестові замовлення (10426–10428). У реальній роботі такі функції тестують на копії бази (тестовому середовищі), а не на робочій.